File: egottol/engines/eii/detectors.py
"""Impulse extraction: continuous waveforms → discrete events."""

from __future__ import annotations

from typing import List

import numpy as np

from egottol.engines.eii.types import EIIConfig, EIIState, ImpulseEvent


class ImpulseDetector:
    """Configurable detector for threshold, differentiator, comparator, RF, memristor modes."""

    MODES = ("threshold", "differentiator", "comparator", "rf_envelope", "memristor_switch")

    def __init__(self, config: EIIConfig):
        self.config = config
        self.threshold = 0.5
        self.tau_d = 1e-3
        self.theta_rf = 0.1
        self.i_set = 1e-6

    def detect(self, state: EIIState, dt: float) -> List[ImpulseEvent]:
        mode = self.config.detector_mode
        if mode not in self.MODES:
            raise ValueError(f"Unknown detector mode: {mode}")

        dispatch = {
            "threshold": self._threshold,
            "differentiator": self._differentiator,
            "comparator": self._comparator,
            "rf_envelope": self._rf_envelope,
            "memristor_switch": self._memristor_switch,
        }
        return dispatch[mode](state, dt)

    def _threshold(self, state: EIIState, dt: float) -> List[ImpulseEvent]:
        events: List[ImpulseEvent] = []
        n = min(len(state.voltages), self.config.num_channels)
        for ch in range(n):
            if state.refractory[ch] > 0:
                state.refractory[ch] = max(0.0, state.refractory[ch] - dt)
                continue
            v = state.voltages[ch]
            if v >= self.threshold:
                events.append(
                    ImpulseEvent(
                        t=state.t,
                        channel=ch,
                        event_type="threshold_cross",
                        amplitude=v,
                    )
                )
                state.refractory[ch] = self.config.window_T * 0.1
        return events

    def _differentiator(self, state: EIIState, dt: float) -> List[ImpulseEvent]:
        events: List[ImpulseEvent] = []
        n = min(len(state.voltages), self.config.num_channels)
        for ch in range(n):
            dv = (state.voltages[ch] - state.prev_voltages[ch]) / max(dt, 1e-12)
            a = self.tau_d * dv + state.voltages[ch]
            prev_a = state.prev_diff_signal[ch]
            if a > self.threshold and a < prev_a:
                events.append(
                    ImpulseEvent(
                        t=state.t,
                        channel=ch,
                        event_type="spike",
                        amplitude=a,
                    )
                )
            state.prev_diff_signal[ch] = a
        state.prev_voltages = state.voltages.copy()
        return events

    def _comparator(self, state: EIIState, dt: float) -> List[ImpulseEvent]:
        events: List[ImpulseEvent] = []
        n = min(len(state.voltages), self.config.num_channels)
        for ch in range(n):
            high = state.voltages[ch] >= self.threshold
            was_high = state.prev_comparator[ch] >= 0.5
            if high and not was_high:
                events.append(
                    ImpulseEvent(
                        t=state.t,
                        channel=ch,
                        event_type="spike",
                        amplitude=state.voltages[ch],
                    )
                )
            state.prev_comparator[ch] = 1.0 if high else 0.0
        return events

    def _rf_envelope(self, state: EIIState, dt: float) -> List[ImpulseEvent]:
        events: List[ImpulseEvent] = []
        alpha = dt / max(self.tau_d, dt)
        n = min(len(state.voltages), self.config.num_channels)
        for ch in range(n):
            target = state.voltages[ch] ** 2
            env = (1 - alpha) * state.prev_envelope[ch] + alpha * target
            if env > self.theta_rf and env < state.prev_envelope[ch]:
                events.append(
                    ImpulseEvent(
                        t=state.t,
                        channel=ch,
                        event_type="rf_burst_peak",
                        amplitude=np.sqrt(env),
                    )
                )
            state.prev_envelope[ch] = env
        return events

    def _memristor_switch(self, state: EIIState, dt: float) -> List[ImpulseEvent]:
        events: List[ImpulseEvent] = []
        n = min(len(state.currents), self.config.num_channels)
        for ch in range(n):
            i_now = state.currents[ch]
            i_prev = state.prev_memristor_current[ch]
            if abs(i_now) > self.i_set and np.sign(i_now) != np.sign(i_prev):
                events.append(
                    ImpulseEvent(
                        t=state.t,
                        channel=ch,
                        event_type="memristor_switch",
                        amplitude=i_now,
                    )
                )
            state.prev_memristor_current[ch] = i_now
        return events

File: core/eii/detectors.cpp
#include "detectors.h"

#include <algorithm>
#include <cmath>

namespace deepiri {

namespace {
double sign(double v) {
    if (v > 0.0) return 1.0;
    if (v < 0.0) return -1.0;
    return 0.0;
}
}  // namespace

std::vector<ImpulseEvent> ImpulseDetector::detect(EIIState& state, double dt) {
    switch (config_.detectorMode) {
        case DetectorMode::Threshold:
            return detectThreshold(state, dt);
        case DetectorMode::Differentiator:
            return detectDifferentiator(state, dt);
        case DetectorMode::Comparator:
            return detectComparator(state, dt);
        case DetectorMode::RfEnvelope:
            return detectRfEnvelope(state, dt);
        case DetectorMode::MemristorSwitch:
            return detectMemristorSwitch(state, dt);
    }
    return {};
}

std::vector<ImpulseEvent> ImpulseDetector::detectThreshold(EIIState& state, double dt) {
    std::vector<ImpulseEvent> events;
    size_t n = std::min(state.voltages.size(), static_cast<size_t>(config_.numChannels));
    for (size_t ch = 0; ch < n; ++ch) {
        if (state.refractory[ch] > 0.0) {
            state.refractory[ch] = std::max(0.0, state.refractory[ch] - dt);
            continue;
        }
        double v = state.voltages[ch];
        if (v >= threshold) {
            events.push_back({state.t, static_cast<int>(ch), "threshold_cross", v});
            state.refractory[ch] = config_.windowT * 0.1;
        }
    }
    return events;
}

std::vector<ImpulseEvent> ImpulseDetector::detectDifferentiator(EIIState& state, double dt) {
    std::vector<ImpulseEvent> events;
    size_t n = std::min(state.voltages.size(), static_cast<size_t>(config_.numChannels));
    for (size_t ch = 0; ch < n; ++ch) {
        double dv = (state.voltages[ch] - state.prevVoltages[ch]) / std::max(dt, 1e-12);
        double a = tauD * dv + state.voltages[ch];
        double prevA = state.prevDiffSignal[ch];
        if (a > threshold && a < prevA) {
            events.push_back({state.t, static_cast<int>(ch), "spike", a});
        }
        state.prevDiffSignal[ch] = a;
    }
    state.prevVoltages = state.voltages;
    return events;
}

std::vector<ImpulseEvent> ImpulseDetector::detectComparator(EIIState& state, double dt) {
    (void)dt;
    std::vector<ImpulseEvent> events;
    size_t n = std::min(state.voltages.size(), static_cast<size_t>(config_.numChannels));
    for (size_t ch = 0; ch < n; ++ch) {
        bool high = state.voltages[ch] >= threshold;
        bool wasHigh = state.prevComparator[ch] >= 0.5;
        if (high && !wasHigh) {
            events.push_back({state.t, static_cast<int>(ch), "spike", state.voltages[ch]});
        }
        state.prevComparator[ch] = high ? 1.0 : 0.0;
    }
    return events;
}

std::vector<ImpulseEvent> ImpulseDetector::detectRfEnvelope(EIIState& state, double dt) {
    std::vector<ImpulseEvent> events;
    double alpha = dt / std::max(tauD, dt);
    size_t n = std::min(state.voltages.size(), static_cast<size_t>(config_.numChannels));
    for (size_t ch = 0; ch < n; ++ch) {
        double target = state.voltages[ch] * state.voltages[ch];
        double env = (1.0 - alpha) * state.prevEnvelope[ch] + alpha * target;
        if (env > thetaRf && env < state.prevEnvelope[ch]) {
            events.push_back({state.t, static_cast<int>(ch), "rf_burst_peak", std::sqrt(env)});
        }
        state.prevEnvelope[ch] = env;
    }
    return events;
}

std::vector<ImpulseEvent> ImpulseDetector::detectMemristorSwitch(EIIState& state, double dt) {
    (void)dt;
    std::vector<ImpulseEvent> events;
    size_t n = std::min(state.currents.size(), static_cast<size_t>(config_.numChannels));
    for (size_t ch = 0; ch < n; ++ch) {
        double iNow = state.currents[ch];
        double iPrev = state.prevMemristorCurrent[ch];
        if (std::abs(iNow) > iSet && sign(iNow) != sign(iPrev)) {
            events.push_back({state.t, static_cast<int>(ch), "memristor_switch", iNow});
        }
        state.prevMemristorCurrent[ch] = iNow;
    }
    return events;
}

}  // namespace deepiri
